reconcile/openshift_rolebindings.py

```diff
@@ -19,40 +31,194 @@
 from reconcile.utils.semver_helper import make_semver
 from reconcile.utils.sharding import is_in_shard
 
-ROLES_QUERY = """
-{
-  roles: roles_v1 {
-    name
-    users {
-      org_username
-      github_username
-    }
-    bots {
-      openshift_serviceaccount
-    }
-    access {
-      namespace {
-        name
-        clusterAdmin
-        managedRoles
-        delete
-        cluster {
-          name
-          auth {
-            service
-          }
+QONTRACT_INTEGRATION = "openshift-rolebindings"
+COMMON_CLUSTER_ROLES = ["view", "edit", "admin", "tekton-trigger-access"]
+QONTRACT_INTEGRATION_VERSION = make_semver(0, 3, 0)
+
+
+class OCResource(BaseModel):
+    resource: OR
+    resource_name: str
+    privileged: bool
+
+    class Config:
+        arbitrary_types_allowed = True
+
+
+@dataclass
+class ServiceAccountSpec:
+    sa_namespace_name: str
+    sa_name: str
+
+    @staticmethod
+    def create_sa_spec(bots: list[BotV1] | None) -> list["ServiceAccountSpec"]:
+        return [
+            ServiceAccountSpec(
+                sa_namespace_name=full_service_account[0],
+                sa_name=full_service_account[1],
+            )
+            for bot in bots or []
+            if bot.openshift_serviceaccount
+            and (full_service_account := bot.openshift_serviceaccount.split("/"))
+            and len(full_service_account) == 2
+        ]
+
+
+class RoleBindingSpec(BaseModel):
+    role_name: str
+    role_kind: str
+    namespace: NamespaceV1
+    cluster: ClusterV1
+    privileged: bool
+    username_list: set[str]
+    openshift_service_accounts: list[ServiceAccountSpec]
+
+    class Config:
+        arbitrary_types_allowed = True
+
+    def get_users_desired_state(self) -> list[dict[str, str]]:
+        return [
+            {"cluster": self.cluster.name, "user": username}
+            for username in self.username_list
+        ]
+
+    @staticmethod
+    def create_role_binding_spec(
+        access: AccessV1,
+        oc_map: ob.ClusterMap | None = None,
+        users: list[UserV1] | None = None,
+        enforced_user_keys: list[str] | None = None,
+        bots: list[BotV1] | None = None,
+    ) -> Optional["RoleBindingSpec"]:
+        if not access.namespace:
+            return None
+        if not (access.role or access.cluster_role):
+            return None
+        privileged = access.namespace.cluster_admin or False
+        if oc_map and not oc_map.get(access.namespace.cluster.name):
+            return None
+        auth_dict = [auth.dict(by_alias=True) for auth in access.namespace.cluster.auth]
+        username_list = RoleBindingSpec.get_usernames_from_role(
+            users,
+            ob.determine_user_keys_for_access(
+                access.namespace.cluster.name,
+                auth_dict,
+                enforced_user_keys,
+            ),
+        )
```

Comment:
the method name is confusing, it's called `get_usernames_from_role` but params are users and user_keys.